keps/582-preempt-based-on-flavor-order/README.md

```diff
@@ -0,0 +1,330 @@
+  next resource flavor when scheduling
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a Kueue administrator I want to ensure more important jobs running on more 
+stable resources. This can happen in case that there are normal and spot instances
+in my cluster. In this case I prefer my high priority jobs not running on spot 
+instances. If high priority jobs can preempt jobs in normal ecs before try spot instances,
+stability can be achieved.
+
+My use case can be supported by set `flavorFungibility` to `BeforeNextFlavor`  in the Kueue configuration.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Cluster Queue API
+
+We extend the Cluster Queue API to introduce the new fields: flavorFungibility to opt-in and configure the new behavior.
+
+```
+type FlavorFungibility string
+
+const (
+	BeforeNextFlavor FlavorFungibility = "BeforeNextFlavor"
+	AfterAllFlavors  FlavorFungibility = "AfterAllFlavors"
+)
+
+// ClusterQueuePreemption contains policies to preempt Workloads from this
+// ClusterQueue or the ClusterQueue's cohort.
+type ClusterQueuePreemption struct {
+	...
+
+	// +kubebuilder:default="AfterAllFlavors"
+	// +kubebuilder:validation:Enum=BeforeNextFlavor;AfterAllFlavors
+	FlavorFungibility FlavorFungibility `json:"flavorFungibility"`
```

Comment:
I agree with the location of the new field, but I think the current name proposals are too obscure:

```yaml
preemption:
  flavorFungibility: BeforeNextFlavor
```

Do the words in this order give some idea of what is going to happen? I don't think they do.

Another thing to consider is whether the feature makes sense even if preemption is disabled, for example, by simply borrowing. Today the algorithm avoids borrowing if it would fit in the next flavor.
Should the policy for borrowing be the same as preemption?

Maybe, maybe not. WDYT?

Perhaps it should be the other way around:

```yaml
flavorFungibility:
  whenCanPreempt: Preempt | TryNextFlavor
  whenCanBorrow: Borrow | TryNextFlavor
```

Reply:
cc @mwielgus 
for thoughts on API

Reply:
+1 for @acondor proposal ("Preempt & Borrow")

Reply:
I love the thinking :)
I had thought of something similar before, but standard questions about the APIs remain:

- How intuitive is it to users?
- How extensible can it be? Do we imagine other modes of fungibility in the future?
- Can we code this in a maintainable way?

Reply:
> How intuitive is it to users?

Here is where real word use cases can inform our decisions. Please share if you have any.

Reply:
> * How intuitive is it to users?

The api maybe like this:
```
type FlavorFungibility string

const (
	Borrow FlavorFungibility = "Borrow"
	Preempt  FlavorFungibility = "Preempt"
        TryNextFlavor FlavorFungibility = "TryNextFlavor"
)
// ClusterQueuePreemption contains policies to preempt Workloads from this
// ClusterQueue or the ClusterQueue's cohort.
type ClusterQueuePreemption struct {
  ...
  // +kubebuilder:validation:UniqueItems=true
  // +kubebuilder:validation:Enum="Borrow,Preempt,TryNextFlavor"
  FlavorFungibility []FlavorFungibility `json:"flavorFungibility"`
}
```

First, I think a list can cover more usage scenarios. It can not only define which phase should take effect, but also define the order of these phase.
Then, if we support more fungibility, we can just add an option to the enum.

For example, it a cluster queue should not borrow before all flavors were considered, then we can write like this:
```
flavorFungibility:
  - preempt
  - tryNextFlavor
  - borrow
```
And if we want to preempt after all flavors were considered, we can write like this:
```
flavorFungibility:
  - borrow
  - tryNextFlavor
  - preempt
```
Or if we want to preempt current flavor if borrowing still not work, we can write like this:
```
flavorFungibility:
  - borrow
  - preempt
  - tryNextFlavor
```

Reply:
What if the user removes `Borrow` or `Preempt`, but has cohorts and a preemption policy. Is borrowing or preemption disabled in this case? Could that be confusing?

Then I would like to see how this can be implemented. Is it a state machine? Do we program all the possible combinations?

I don't have the answers at this moment. Could you write some thoughts in the KEP?
Maybe list all the high level proposals we have already discussed in this thread and show the advantages or disadvantages.

In general I think it's useful to have these policies, we just need to be careful about the API (for users' sake) and implementation (so that the code doesn't end up like spaghetti)